Support `X509Certificate` in INI and HIA request

vnovitskyi opened the pull request:
This PR adds support for using x509 self-signed certificates in INI and HIA requests.
_Details_
- Enables the use of x509 self-signed certificates when handling INI and HIA requests.
- Updates the generation of `init_letter` to use certificate data instead of a,x and e keys.
- Adds support for passing a debug_mode: boolean option to client initialization, allowing easy logging of requests and responses for debugging purposes. When debug_mode is enabled, Faraday’s logger middleware is activated to output request and response bodies to STDOUT

_Impact_
- No breaking changes; existing workflows not using certificates remain unaffected.
- Users can now configure and use certificates for INI and HIA interactions as needed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/railslove/epics?pullRequest=168) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/railslove/epics?pullRequest=168) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/railslove/epics?pullRequest=168) it.</sub>

tobischo commented on lib/epics/client.rb:
```diff
@@ -1,21 +1,27 @@
+    self.locale = options[:locale] || Epics::DEFAULT_LOCALE
+    self.product_name = options[:product_name] || Epics::DEFAULT_PRODUCT_NAME
+    self.debug_mode = !!options[:debug_mode]
+    self.x_509_certificate_a_content = options[:x_509_certificate_a_content]
+    self.x_509_certificate_x_content = options[:x_509_certificate_x_content]
+    self.x_509_certificate_e_content = options[:x_509_certificate_e_content]
```
Is there any case where only one of them should be passed?

vnovitskyi replied:
Theoretically, yes—because each request type requires a different x509 certificate. For example, if you only need to make an `INI` request, you just need `x_509_certificate_a_content`. For an `HIA` request, you’ll need both `x_509_certificate_x_content` and `x_509_certificate_y_content`.

tobischo commented on lib/epics/ini.rb:
```diff
@@ -23,9 +23,20 @@ def body
     Nokogiri::XML::Builder.new do |xml|
       xml.SignaturePubKeyOrderData('xmlns:ds' => 'http://www.w3.org/2000/09/xmldsig#', 'xmlns' => 'http://www.ebics.org/S001') {
         xml.SignaturePubKeyInfo {
+          if x_509_certificate_a
```
The x509 xml data looks the same in all 3 places, just using a different input.

I would suggest moving that into a function/method to re-use and shorten the part of the xml document.

tobischo commented on lib/epics/client.rb:
```diff
@@ -1,21 +1,27 @@
-  attr_accessor :passphrase, :url, :host_id, :user_id, :partner_id, :keys, :keys_content, :locale, :product_name
-  attr_writer :iban, :bic, :name
+  attr_accessor :passphrase, :url, :host_id, :user_id, :partner_id, :keys, :keys_content, :locale, :product_name,
+                :x_509_certificate_a_content, :x_509_certificate_x_content, :x_509_certificate_e_content, :debug_mode
```
Having individual long method names for each individual certificate does not seem very elegant.

Having the 3 keys combined under 1 method allows for keeping the immediate surface area of the client class way smaller.

tobischo reviewed (changes requested):
This looks generally ok. I left some comments for which I would suggest a bit of refactoring. I have no EBICS access with which I could test it.

In case where certificates are being used, it should either be all or nothing. Therefore we should probably validate for that.

We probably also do not need to keep all 3 as attr_accessor variables on the client.

The most important piece that is missing here are usage examples for the README.
Without that, the change is not really complete for something as specific as this.
I would not expect anyone to figure it out by reading the code unless they know that it is already an available feature.

tobischo reviewed (approved):
LGTM, also tested at least file download on EBICS 2.5 still working (the primary use case that I have and can easily test)

The primary thing that is keeping me from merging this right now is #154, as bringing those together again might be a larger change and change the interface

vnovitskyi commented:
> LGTM, also tested at least file download on EBICS 2.5 still working (the primary use case that I have and can easily test)
> 
> The primary thing that is keeping me from merging this right now is #154, as bringing those together again might be a larger change and change the interface

Hi @tobischo 
I'm happy to see that the PR has been approved! I also added a final commit updating the documentation on how to use X.509 certificates.
If you require any additional assistance with this topic, please don't hesitate to let me know. And apologies that addressing the comments took a bit longer than expected.
